Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -1,443 +1,381 @@
+    NSInteger maxOps = self.maxConcurrentOperationCount;
+    return (maxOps <= 0) || (_currentConcurrentOperationCount < maxOps);
 }
 
-NSObject* NSAtomicListPop(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* node = *listPtr;
-    if (!node) {
-        return NULL;
+// Private helper that tries to starts a single NSOperation on the dispatch queue and manages associated state.
+// If the operation is not startable due to the state of the queue (ie: suspension, above max concurrent operations)
+// it is enqueued for later.
+// If the operation is not ready, this is a no-op
+// (the operation can try to start again later once the queue receives a KVO notification that it became ready)
+- (void)_startOperation:(NSOperation*)operation {
+    if (operation.isExecuting || operation.isFinished) {
+        TraceError(g_logTag, L"Attempted to start an operation that was already executing or finished; ignoring.");
+        return;
     }
 
-    *listPtr = node->next;
+    if (!operation.isReady) {
+        // Deliberate no-op (see function comments)
+        return;
+    }
 
-    NSObject* elt = [[node->elt retain] autorelease];
-    delete node;
-    return elt;
-}
+    {
+        std::lock_guard<std::recursive_mutex> lock(_concurrentOperationCountLock);
+        if ((self.isSuspended) || (![self _belowMaxConcurrentOperations])) {
+            // Enqueue the operation for later
+            _queuedOperations.push(operation);
+            return;
+        }
 
-NSObject* NSAtomicListPeek(NSAtomicListRef* listPtr) {
-    struct NSAtomicListNode* node = *listPtr;
-    if (!node) {
-        return NULL;
-    }
+        ++_currentConcurrentOperationCount;
+    } // _concurrentOperationCountLock scope
 
-    return node->elt;
-}
+    std::lock_guard<std::recursive_mutex> lock(_dispatchQueueLock);
+    dispatch_async(_dispatchQueue, ^{
+        // Check if currentQueue needs to be changed
+        NSMutableDictionary* threadDictionary = [[NSThread currentThread] threadDictionary];
+        StrongId<NSOperationQueue> currentQueue = [threadDictionary objectForKey:_NSOperationQueueCurrentQueueKey];
+        bool needToChangeCurrentQueue = (![[NSThread currentThread] isMainThread]) && (![self isEqual:currentQueue]);
 
-void NSAtomicListInsert(NSAtomicListRef* listPtr, NSObject* elt) {
-    struct NSAtomicListNode* node = new NSAtomicListNode();
-    node->elt = elt;
+        // If so, cache the previous currentQueue stored in the threadDictionary, and store the new currentQueue
+        if (needToChangeCurrentQueue) {
+            [threadDictionary setObject:self forKey:_NSOperationQueueCurrentQueueKey];
+        }
 
-    do {
-        node->next = *listPtr;
-    } while (CompareExchangePointer((void**)listPtr, node, node->next) != node->next);
-}
+        // Restore state once the operation is done, and attempt to start any queued operations
+        auto cleanup = wil::ScopeExit([self, threadDictionary, currentQueue, needToChangeCurrentQueue]() {
```

Comment:
no wait i didn't think this all the way through.

- isFinished is set at the end of [operation start]. if we schedule more operations when we observe isFinished, we end up doing so before we decrement _currentConcurrentOperationCount, causing _popAndStart to do the wrong thing.
- we can move the decrement to the isFinished observation, but i dislike this because 1) the increment and decrement ideally should be in the same function 2) an operation can finish through means besides _startOperation (ie: if a user decides to start it manually for whatever reason)
- as it currently is, we're basically checking if new operations need to be started when an old operation has finished, using the same async context. i think that this makes intuitive sense, so i'm going to leave it as is for now. (it might look less scary once you see the thread_local version also)